kernel/shared/src/main/scala/cats/effect/kernel/Async.scala

```diff
@@ -150,6 +150,40 @@ trait Async[F[_]] extends AsyncPlatform[F] with Sync[F] with Temporal[F] {
   def async_[A](k: (Either[Throwable, A] => Unit) => Unit): F[A] =
     async[A](cb => as(delay(k(cb)), None))
 
+  /**
+   * Like [[Sync.blocking]] but intended for blocking operations that are not interruptible but
+   * can be externally canceled, typically by closing a resource. This should not be used to
+   * implement cancelation via thread interruption; instead use [[Sync.interruptible]] or
+   * [[Sync.interruptibleMany]].
+   *
+   * {{{
+   * Async[F].blockingCancelable(Async[F].blocking(socket.close()))(socket.read(buffer))
+   * }}}
+   *
+   * @param cancel
+   *   An effect that externally cancels the blocking operation.
+   *
+   * @param thunk
+   *   The side effect which is to be suspended in `F[_]` and evaluated on a blocking execution
+   *   context
+   */
+  def blockingCancelable[A](cancel: F[Unit])(thunk: => A): F[A] =
+    flatMap2(deferred[Unit], delay(new AtomicBoolean(false))) { (gate, inProgress) =>
+      map(
+        race(
+          productR(gate.get)( // ensure cancelation fiber started before blocking
+            blocking {
+              inProgress.set(true)
+              try thunk
+              finally inProgress.set(false)
+            }),
+          onCancel(
+            productR(gate.complete(()))(never[A]),
+            ifM(delay(inProgress.get()))(cancel, unit)
+          )
+        ))(_.merge)
+    }
```

Comment:
For those reading the code, could you extract each of the two sides as a variable and add a 1-line comment on each, explaining what each side does, if you please?

```suggestion
      // LHS: Once cancellation fiber started, do blocking operation?
      val lhs = 
          productR(gate.get)( // ensure cancelation fiber started before blocking
            blocking {
              inProgress.set(true)
              try thunk
              finally inProgress.set(false)
            })

      // RHS: wait for a cancellation, if received cancel ??
      val rhs = 
        onCancel(
          productR(gate.complete(()))(never[A]),
          ifM(delay(inProgress.get()))(cancel, unit)
        )

      map(race(lhs, rhs), _.merge)
    }
```